src/jit/compiler.h

```diff
@@ -3861,13 +3861,26 @@ public :
     // Convert a BYTE which represents the VM's CorInfoGCtype to the JIT's var_types
     var_types   getJitGCType(BYTE gcType);
 
-    // Get the "primitive" type, if any, that is used to pass or return
-    // values of the given struct type.
-    var_types    argOrReturnTypeForStruct(CORINFO_CLASS_HANDLE clsHnd, bool forReturn);
+    enum structPassingKind { SPK_Unknown,             // Invalid value, never returned
+                             SPK_PrimitiveType,       // The struct is passed/returned using a primitive type
+                             SPK_ByValue,             // The struct is passed/returned by value, usually in multiple registers
+                             SPK_ByValueAsHfa,        // The struct is passed/returned by value as an HFA in multiple registers
+                             SPK_ByReference };       // The struct is passed/returned by reference to a copy/buffer
+
+    // Get the "primitive" type that is is used when we are given a struct of size 'structSize'
+    // when give a pointer sized struct the clsHnd is used to determine if the struct contains GC ref
+    var_types    getPrimitiveTypeForStruct(unsigned structSize, CORINFO_CLASS_HANDLE clsHnd);
+
+    // Get the type that is used to pass values of the given struct type.
+    // If you already have retrieved the struct size then pass it as the third optional argument
+    //
+    var_types    getArgTypeForStruct(CORINFO_CLASS_HANDLE clsHnd, structPassingKind* wbPassStruct, unsigned structSize = 0);
+
+    // Get the type that is used to return values of the given struct type.
+    // If you already have retrieved the struct size then pass it as the third optional argument
+    //
+    var_types    getReturnTypeForStruct(CORINFO_CLASS_HANDLE clsHnd, structPassingKind* wbPassStruct, unsigned structSize = 0);
 
-    // Slightly optimized version of the above where we've already computed the size,
-    // so as to avoid a repeated JIT/EE interface call.
-    var_types    argOrReturnTypeForStruct(unsigned size, CORINFO_CLASS_HANDLE clsHnd, bool forReturn);
 
```

Comment:
Looks like not all callers need *wbPassStruct filled in, so maybe you should allow a default of nullptr.